### DISCUSSION TITLE
Configure for (MIS) Multi Inverter Support

### DISCUSSION THREAD

**Discussion by fritserasmus** [SBFspot/sbfspot-config#60] [Q&A] [answered]
I have a SBFspot install dated Feb 2018
The "SBFspot.cfg" file has a section "# MIS_Enabled (Multi Inverter Support: Default=0 Disabled)"

0 being 1 Inverter
1 being more than 1

In the file: SBFspotUpload.cfg in  /usr/local/bin/sbfspot.3 the file contains info about the serial number of the two inverters 

This ensured that both inverters data were written to a csv file
e.g:
```
07/02/2023 16:10:04;SB 4000TL-21 237 Left;SB 4000TL-21;2130440300;1297,000;155,000;6,270;0,720;206,960;216,360;1430,000;0,000;0,000;6,147;0,000;0,000;232,680;0,000;0,000;1452,000;1430,000;0,000;20,151;32082,431;50,140;24623,776;23867,666;0,000;Ok;Closed;53,360
07/02/2023 16:35:04;SB 4000TL-21 358 Right;SB 4000TL-21;2130440301;753,000;176,000;3,437;0,673;219,230;262,750;918,000;0,000;0,000;3,973;0,000;0,000;231,450;0,000;0,000;929,000;918,000;0,000;21,498;32840,302;49,930;24718,437;24011,062;69,020;Ok;Closed;48,210
```
This setup now uploads the one inverter data to [https://pvoutput.org/](url)

Since I now need the readings in my MQTT server and decided it is time for a fresh install and followed the instructions provided here.
[https://github.com/SBFspot/sbfspot-config](url)

The issue now is that I have only ONE inverter configured in the  SBFspotUpload.cfg file and there is no indication for the program to know I have two inverters
The SBFspotUpload.cfg file also caters for only one inverter.

My questions are: 
What do I do to configure for two inverters?
How do I amend the SBFspotUpload.cfg file to know I have two inverters

How do I configure MQTT in the SBFspotUpload.cfg under MQTT config section to have two inverters?

Once configured: How do I restart the SBFSpot service?

I will really appreciate help/pointers very much

**Comment by sbf-** [comment]
> The "SBFspot.cfg" file has a section "# MIS_Enabled (Multi Inverter Support: Default=0 Disabled)"

The `#` disables this config line, so by default it's disabled. Add MIS_Enabled=1 to your config

If SBFspot collects the data of 2 inverters (via MIS_Enabled), it will send MQTT data for both

I know sbfspot-config is not bullet-proof, especially not for multi inverter plants as I'm unable to test this setup (only have one SB4000TL-20)

**Reply by fritserasmus** [reply]
Thanks sbf,
I was thinking I should add that section in the new install.
I now have two RPis old one, I built in 2018 and rebuilt 2019 that still writes data in the db and the csv files.


So for me to copy that section that I know works will be fairly easy - I just wanted confirmation that the SBF executable will still honor the MIS setup.

I was able to get ALL the payloads to be read by Node-Red for the primary SB 4000.

I am now ready to experiment with the the config file by adding mutiple inverters.

The part that I am not sure how to address is the MQTT publish section. Currently I have a fixed serial number. Surely it should be a variable that will be read from the config file where multi serials are specified - To figure that out is above my paygrade :-) 

I will make the changes I am comfortable and then check in for help again.

Once I get all this to work I am happy to share my learnings here.

I am aiming to have a dashboard showing my 4 sets of pv modules production, migrated into one feed to my SI, that feeds and pulls from the batteries.

I will also have the big energy users on gauges and 72 hour line graph.

so 
PV -> SB ->SI <-> Battery
PV -> GRID
GRID to DB 
Organised in a picture

**Reply by fritserasmus** [reply]
> > The "SBFspot.cfg" file has a section "# MIS_Enabled (Multi Inverter Support: Default=0 Disabled)"
> 
> The `#` disables this config line, so by default it's disabled. Add MIS_Enabled=1 to your config
> 
> If SBFspot collects the data of 2 inverters (via MIS_Enabled), it will send MQTT data for both
> 
> I know sbfspot-config is not bullet-proof, especially not for multi inverter plants as I'm unable to test this setup (only have one SB4000TL-20)

Thanks @sbf- ,

I have an install that publishes payload of one inverter just perfectly. (Opted for speedwire rather than BT)

In my MQTT Settings I have the following challenge:- (I have masked the serial number)


Original vinilla:-
--MQTT_Topic=sbfspot_{serial}


This one publishes the payload of the primary SunnyBoy to MQTT server perfectly
--MQTT_Topic=sbfspot/213000358

I tried to use this as a variable in the payload hoping that two payloads, one for each inverter will be sent through, but appears it's not working as I was hoping
MQTT_Topic=sbfspot_1856/{InvSerial}

Any suggestions on what to do to be able to get two payloads published - one for each inverter?

**Reply by fritserasmus** [reply]
@sbf- 

So I thought I'd experiment a little further with this:
Please note: I am NOT a developer myself but I am a Project Manager in the software development space.

In the "SBFspot.cfg" file I made provisions for two IP addresses and two MQTT output payloads 
IP address No1 will be for Inverter_LEFT    
IP address No2 will be for Inverter_RIGHT 

Saving the config file with the first combination publish the correct payload for Inverter_LEFT
and  when I save the config file with the second combo, the payload has the second inverter (Inverter_Right) info

Is it possible to change the program to then run:- 
SBFspot_1.cfg
then
SBFspot_2.cfg
then
SBFspot_3.cfg.........etc, 

as that will solve the issue of multiple inverters.

Regards

Frits

**Reply by sbf-** [reply]
> In the "SBFspot.cfg" file I made provisions for two IP addresses and two MQTT output payloads
> IP address No1 will be for Inverter_LEFT
> IP address No2 will be for Inverter_RIGHT

How did you do this? The correct syntax is `IP_Address=192.168.1.10,192.168.1.11`
So, a list of comma separated IP addresses of your inverters.

That way both inverters will expose their data to MQTT

To better understand the problem, post your config file and SBFspot verbose output (-v5)

Another option is to use 2 config files, one for each inverter and call SBFspot twice with `-cfg` option for each config

SBFspot_1.cfg (`IP_Address=192.168.1.10`)
SBFspot_2.cfg (`IP_Address=192.168.1.11`)

Call SBFspot with `-cfgSBFspot_1.cfg` and a second time with `-cfgSBFspot_2.cfg`

**Reply by fritserasmus** [reply]
I think I was not clear in my explanation as I was probably too excited to get the payload through to my MQTT server .
I effectively have two configs in one. So if I in turn comment out the pair that I do not want the data is published for the active pair (Pair = IP address and MQTT topic of choice)
Attached is the config file. I am not sure how to do a "SBFspot verbose output (-v5)"

```
################################################################################
#                     ____  ____  _____                _                        
#                    / ___|| __ )|  ___|__ _ __   ___ | |_                      
#                    \___ \|  _ \| |_ / __| '_ \ / _ \| __|                     
#                     ___) | |_) |  _|\__ \ |_) | (_) | |_                      
#                    |____/|____/|_|  |___/ .__/ \___/ \__|                     
#                                         |_|                                   
#                                                                               
#  SBFspot.cfg - Configuration file for SBFspot.exe                             
#  SBFspot - Yet another tool to read power production of SMA solar inverters  
#  (c)2012-2023, SBF (https://github.com/SBFspot/SBFspot)              
#                                                                               
#  DISCLAIMER:                                                                  
#  A user of SBFspot software acknowledges that he or she is receiving this     
#  software on an "as is" basis and the user is not relying on the accuracy   
#  or functionality of the software for any purpose. The user further           
#  acknowledges that any use of this software will be at his own risk           
#  and the copyright owner accepts no responsibility whatsoever arising from    
#  the use or application of the software.                                      
#                                                                               
#    SMA and Speedwire are registered trademarks of SMA Solar Technology AG     
################################################################################
#
# SBFspot.cfg - Created Thu 02 Mar 2023 10:42:15 PM SAST with sbfspot-config V1.3.4
#

# SMA Inverter's Speedwire IP address
# If IP_Address is set to 0.0.0.0 SBFspot will auto-detect connected inverters
# If IP_Address is set one or more valid IP's, SBFspot will connect directly to these IP's
# Multiple IP addresses can be provided (comma separated)
# IP_Address=192.168.1.22
#   UNCOMMENT die TWEE saam: IP_Address en MQTT_Topic

# SunnyBoy Inverters - 
################################

# Left - 21
#*********************
#IP_Address=192.168.1.21
#MQTT_Topic=sbfspot/2130438237

#Right -  22
#*********************
IP_Address=192.168.1.22
MQTT_Topic=sbfspot/2130438358
################################


# User password (default 0000)
Password=****

# Plantname
Plantname=Frits1856

# OutputPath (Place to store CSV files)
# Windows: C:\Users\Public\SMAdata\%Y
# Linux  : /home/pi/smadata/%Y
# %Y %m and %d will be expanded to Year Month and Day
OutputPath=/home/pi/smadata/%Y

# OutputPathEvents (Place to store CSV files for events)
# If omitted, OutputPath is used
OutputPathEvents=/home/pi/smadata/%Y/Events

# Timezone
# Select the right timezone in date_time_zonespec.csv
# e.g. Timezone=Europe/Brussels
Timezone=Africa/Johannesburg

# Position of pv-plant https://www.gps-coordinates.net/maps
Latitude=-26.20
Longitude=28.05

# Locale
# Translate Entries in CSV files
# Supported locales: de-DE;en-US;fr-FR;nl-NL;es-ES;it-IT
# Default en-US
Locale=en-US

# DateTimeFormat (default %d/%m/%Y %H:%M:%S)
# For details see strftime() function
# http://www.cplusplus.com/reference/clibrary/ctime/strftime/
DateTimeFormat=%d/%m/%Y %H:%M:%S

# DateFormat (default %d/%m/%Y)
DateFormat=%d/%m/%Y

# TimeFormat (default %H:%M:%S)
TimeFormat=%H:%M:%S

# DecimalPoint (comma/point default comma)
DecimalPoint=comma

# Calculate Missing SpotValues
# If set to 1, values not provided by inverter will be calculated
# eg: Pdc1 = Idc1 * Udc1
CalculateMissingSpotValues=0

# SynchTime (0-30 - 0=disabled, 1=once a day (default), 7=once a week, 30=once a month)
# If set to non-zero value, the plant time is synchronised with local host time
# Some inverters don't have a real-time clock
SynchTime=1

# SynchTimeLow (1-120 - default 1)
# SynchTimeHigh (1200-3600 - default 3600)
# Plant time is adjusted to local host time when SynchTime=1 and
# time difference is between SynchTimeLow and SynchTimeHigh limits
SynchTimeLow=60
SynchTimeHigh=3600

# SunRSOffset
# Offset to start before sunrise and end after sunset (0-3600 - default 900 seconds)
SunRSOffset=3600

###########################
### CSV Export Settings ###
###########################

# With CSV_* settings you can define the CSV file format

# CSV_Export (default 1 = Enabled)
# Enables or disables the CSV Export functionality
CSV_Export=1

# CSV_ExtendedHeader (default 1 = On)
# Enables or disables the SMA extended header info (8 lines)
# isep=;
# Version CSV1|Tool SBFspot|Linebreaks CR/LF|Delimiter semicolon|Decimalpoint comma|Precision 3
# etc...
# This is usefull for manual data upload to pvoutput.org
CSV_ExtendedHeader=1

# CSV_Header (default 1 = On)
# Enables or disables the CSV data header info (1 line)
# dd/MM/yyyy HH:mm:ss;kWh;kW
# This is usefull for manual data upload to pvoutput.org
# If CSV_ExtendedHeader is enabled, CSV_Header is also enabled
CSV_Header=1

# CSV_SaveZeroPower (default 1 = On)
# When enabled, daily csv files contain all data from 00:00 to 23:55
# This is usefull for manual data upload to pvoutput.org
CSV_SaveZeroPower=1

# CSV_Delimiter (comma/semicolon default semicolon)
CSV_Delimiter=semicolon

# CSV_Spot_TimeSource (Inverter|Computer default Inverter)
CSV_Spot_TimeSource=inverter

# CSV_Spot_WebboxHeader (Default 0 = Off)
# When enabled, use Webbox style header (DcMs.Watt[A];DcMs.Watt[B]...)
CSV_Spot_WebboxHeader=0

###########################
###   SQL DB Settings   ###
###########################

SQL_Database=/home/pi/smadata/SBFspot.db

#########################
###   MQTT Settings   ###
#########################

# Full path to mosquitto_pub executable
MQTT_Publisher=/usr/bin/mosquitto_pub

# IP address or DNS name of MQTT Broker
# Don't use test broker for production environments
#MQTT_Host=test.mosquitto.org
MQTT_Host=192.168.1.110

# Port (Optional)
MQTT_Port=1883

# Topic Name
# When using public broker like test.mosquitto.org make sure to define a unique topic name
# by appending the serial number of the inverter to avoid you receive data of another SBFspot user.
# MQTT_Topic=sbfspot_{serial}
# MQTT_Topic=sbfspot_1856/{InvSerial}

# Kyk BO-AAN die leer vir die twee PAYLOADS vir elke SB appart
# 21
# MQTT_Topic=sbfspot/2130438237


# 22
# MQTT_Topic=sbfspot/2130438358


# Format of message items to be sent
# JSON: MQTT_ItemFormat="{key}": {value}
# TEXT: MQTT_ItemFormat={key}:{value}
# XML:  MQTT_ItemFormat=<item name="{key}" value="{value}" />
MQTT_ItemFormat="{key}": {value}

# MQTT_ItemDelimiter (none|blank|comma|semicolon default comma)
# JSON: MQTT_ItemDelimiter=comma
# TEXT: MQTT_ItemDelimiter=semicolon
# XML : MQTT_ItemDelimiter=none
MQTT_ItemDelimiter=comma

# Mandatory arguments for mosquitto_pub executable
# JSON: MQTT_PublisherArgs=-h {host} -t {topic} -m "{{message}}"
# TEXT: MQTT_PublisherArgs=-h {host} -t {topic} -m "{message}"
# XML : MQTT_PublisherArgs=-h {host} -t {topic} -m "<mqtt_message>{message}</mqtt_message>"
MQTT_PublisherArgs=-u erasmi -P 1856 -h {host} -t {topic} -m "{{message}}"

# Data to be published (comma delimited)
MQTT_Data=Timestamp,SunRise,SunSet,InvSerial,InvName,InvTime,InvStatus,InvTemperature,InvGridRelay,EToday,ETotal,PACTot,UDC1,UDC2,IDC1,IDC2,PDC1,PDC2

# Possible keywords are (if supported by your inverter):
# SBFspot Alias   Code                Description
# =======================================================================
# Timestamp                           Current date/time
# InvTime                             Inverter date/time
# SunRise                             Sunrise time
# SunSet                              Sunset time
# InvSerial                           Serial Number
# InvName         NameplateLocation   Device name
# InvClass        NameplateMainModel  Device class
# InvType         NameplateModel      Device type
# InvSwVer        NameplatePkgRev     Software package
# InvStatus       OperationHealth     Condition
# InvTemperature  CoolsysTmpNom       Operating condition temperature
# InvGridRelay    OperationGriSwStt   Grid relay/contactor
# ETotal          MeteringTotWhOut    Total yield
# EToday          MeteringDyWhOut     Day yield
# PACTot          GridMsTotW          Power
# PDC1/PDC2       DcMsWatt            DC power input String 1/2
# UDC1/UDC2       DcMsVol             DC voltage input String 1/2
# IDC1/IDC2       DcMsAmp             DC current input String 1/2
# OperTm          MeteringTotOpTms    Operating time
# FeedTm          MeteringTotFeedTms  Feed-in time
# PAC1            GridMsWphsA         Power L1
# PAC2            GridMsWphsB         Power L2
# PAC3            GridMsWphsC         Power L3
# UAC1            GridMsPhVphsA       Grid voltage phase L1
# UAC2            GridMsPhVphsB       Grid voltage phase L2
# UAC3            GridMsPhVphsC       Grid voltage phase L3
# IAC1            GridMsAphsA_1       Grid current phase L1
# IAC2            GridMsAphsB_1       Grid current phase L2
# IAC3            GridMsAphsC_1       Grid current phase L3
# GridFreq        GridMsHz            Grid frequency
# BatTmpVal       BatTmpVal           Battery temperature
# BatVol          BatVol              Battery voltage
# BatAmp          BatAmp              Battery current
# BatChaStt       BatChaStt           Current battery charge status

# End of Config
```

**Reply by sbf-** [reply, marked as answer]
You should have this in your config:

```
# SunnyBoy Inverters - 
################################

# Left - 21 / Right - 22
# **********************
IP_Address=192.168.1.21,192.168.1.22
MQTT_Topic=sbfspot/{serial}

################################
```

**Reply by fritserasmus** [reply]
Thanks, @sbf- ,
No sun at the moment (20h02) here
I will make the changes and try tomorrow.

Thank you very much for your help!!

**Reply by fritserasmus** [reply]
Please explain to me how I should have pasted the config file in the box so it looks like now.

**Reply by sbf-** [reply]
> Please explain to me how I should have pasted the config file in the box so it looks like now

Use 3 backticks to start and end the block of code

[Markdown doc](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax)

**Reply by fritserasmus** [reply]
@sbf- Thank you very much for your help.

I receive all the readings on my MQTT server just fine.
The only thing I would like to change is the frequency of the payloads sent to the MQTT server.
5 minutes is a huge delay if I compare values with my sunnyportal/sunnyexplorer and if I want to use these values to control contactors to the heavy power use equipment in my house. If it is cloudy and sun comes and goes, success of the usage of MQTT must have a higher frequency of payload  messages

1. If I stop sending values to pvoutput using this pi and ONLY use it for MQTT feeds, how can I increase the publish rate to say every 3 seconds? 

2. What would be my BEST way of stopping feeds to pvoutput?


In case you are interested in how you helped me!!!
Below my dashboard currently in Node-RED.
Two inverters - on the wall left and right. Each has connected to it:- Two panes. One facing East and one facing west.

So I needed to see Left & Right inverter total  
as well as East & West totals

![image](https://user-images.githubusercontent.com/7281074/224662782-705e4857-bb3f-4d67-bd4b-48e4d0f8ff22.png)

Sometimes a reading does come through on the payload like this
![image](https://user-images.githubusercontent.com/7281074/224664048-c17bdca0-f646-42c4-b2d8-c4db22291a9c.png)

I still have to figure out how to keep payloads if nothing new is received - but 5 minutes is a LONG period

Aerial view
![image](https://user-images.githubusercontent.com/7281074/224663598-6c7c8add-c7ce-4bab-9469-7727ec0a0e8b.png)
